README.txt
= statsample-sem

* http://ruby-statsample.rubyforge.org/

== DESCRIPTION:

Structural equation modeling (SEM) for statsample gem, usign ruby and R

== FEATURES/PROBLEMS:

* Abstract generation of models. You could use R packages sem or OpenMx only changing the configuration for engine
* Generates visual representations of output models using GraphViz (to implement)

== SYNOPSIS:

    require 'statsample'
    require 'statsample/sem'
    require 'matrix'
    
    
    matrix=Matrix[[0.1985443,0.1999953,0.2311884,0.2783865,0.3155943],
    [0.1999953,0.2916950,0.2924566,0.3515298,0.4019234],
    [0.2311884,0.2924566,0.3740354,0.4061291,0.4573587],
    [0.2783865,0.3515298,0.4061291,0.5332788,0.5610769],
    [0.3155943,0.4019234,0.4573587,0.5610769,0.6703023]]
    
    
    cases=500
    sem1=Statsample::SEM.new(:name=>"SEM analysis") do |m|
    m.data_from_matrix(matrix,:cases=>cases)
    m.manifests m.data_variables
    m.latents %w{G}
    m.path :from=>m.latents, :to=>m.manifests
    m.path :from=>m.manifests
    m.path :from=>m.latents, :free=>false, :values=>1.0
    
    end
    sem1.compute
    puts sem1.summary
    
    sem2=sem1.dup
    sem2.make_null
    
    sem2.name="Null Model"
    
    sem2.compute
    
    puts sem2.summary

=== OUTPUT 

    = SEM analysis
      Manifests: VAR1, VAR2, VAR3, VAR4, VAR5
      Latents  : G
      Chi-square: 7.384 (d.f=5)
      Parameter estimation
    +------+------+---------------+------------+------------+
    | From |  To  |     Label     |  estimate  |     de     |
    +------+------+---------------+------------+------------+
    | G    | G    | var G (Fixed) | 1.0        | --         |
    | G    | VAR1 | G to VAR1     | 0.39715249 | 0.01558258 |
    | G    | VAR2 | G to VAR2     | 0.50366170 | 0.01827672 |
    | G    | VAR3 | G to VAR3     | 0.57724183 | 0.02049979 |
    | G    | VAR4 | G to VAR4     | 0.70277431 | 0.02407548 |
    | G    | VAR5 | G to VAR5     | 0.79625084 | 0.02674307 |
    | VAR1 | VAR1 | var VAR1      | 0.04081422 | 0.00282540 |
    | VAR2 | VAR2 | var VAR2      | 0.03801998 | 0.00281854 |
    | VAR3 | VAR3 | var VAR3      | 0.04082727 | 0.00316526 |
    | VAR4 | VAR4 | var VAR4      | 0.03938699 | 0.00342167 |
    | VAR5 | VAR5 | var VAR5      | 0.03628702 | 0.00369160 |
    +------+------+---------------+------------+------------+
    
    = Null Model
      Manifests: VAR1, VAR2, VAR3, VAR4, VAR5
      Latents  : G
      Chi-square: 3725.060 (d.f=10)
      Parameter estimation
    +------+------+---------------+------------+------------+
    | From |  To  |     Label     |  estimate  |     de     |
    +------+------+---------------+------------+------------+
    | G    | G    | var G (Fixed) | 1.0        | --         |
    | VAR1 | VAR1 | var VAR1      | 0.19854430 | 0.01258228 |
    | VAR2 | VAR2 | var VAR2      | 0.29169500 | 0.01847955 |
    | VAR3 | VAR3 | var VAR3      | 0.37403540 | 0.02369243 |
    | VAR4 | VAR4 | var VAR4      | 0.53327880 | 0.03377395 |
    | VAR5 | VAR5 | var VAR5      | 0.67030230 | 0.04244876 |
    +------+------+---------------+------------+------------+


== REQUIREMENTS:

* statsample
* R with sem and/or OpenMx packages installed

== INSTALL:

* sudo gem install statsample-sem

== LICENSE:

GPL-2 


History.txt
=== 0.1.0 / 2010-07-03

* First operational version, with 'sem' and 'OpenMx' R's libraries.


spec/spec_helper.rb
# encoding: utf-8

puts "Statsample-SEM specs: Running on Ruby Version: #{RUBY_VERSION}"

require "rubygems"
require 'spec'
require 'spec/autorun'
require 'statsample'
$LOAD_PATH.unshift File.join(File.dirname(__FILE__), '..', 'lib') 
require 'statsample/sem'


spec/openmxengine_spec.rb
$:.unshift(File.dirname(__FILE__)+"/.")
require 'spec_helper'

describe Statsample::SEM::OpenMxEngine do
  before(:all) do
    begin 
      @data_path=File.dirname(__FILE__)+"/fixtures/demo_open_mx.ds"
      @ds=Statsample.load(@data_path)
    rescue 
      @data_path=File.dirname(__FILE__)+"/fixtures/demo_open_mx.csv"
      @ds=Statsample::CSV.read(@data_path)
    end
    @cov_matrix=Statsample::Bivariate.covariance_matrix(@ds)
    @cases=@ds.cases
  end
  describe "using matrix based data" do 
    before(:all) do
      @model=Statsample::SEM::Model.new do |m|
        m.manifests @ds.fields
        m.latents %w{G}
        m.path :from=>m.latents, :to=>m.manifests
        m.path :from=>m.manifests
        m.path :from=>m.latents, :free=>false, :values=>1.0
        m.data_from_matrix(@cov_matrix,:cases=>@cases)
      end
      @engine=Statsample::SEM::OpenMxEngine.new(@model)
      @engine.compute
    end
    it "should generate a valid r query" do
      @engine.r_query.size.should>=0
    end
    it "should have valid std coeffs" do
      sc=@engine.standarized_coefficients
      sc.should be_instance_of Hash
      sc[['G','x1']][:estimate].should be_close(0.891309,0.0001)
    
    end

    it "should return a valid summary" do
      @engine.summary.size>0
    end
    it "should compute and return well formed response" do
      lambda{@engine.compute}.should_not raise_error
      @engine.r_summary.should be_instance_of (Array)
    end
    it "should return a valid graphviz definition for model" do
      @engine.graphviz.size.should>=0
    end
    it "method chi_square return X^2 of model" do 
      @engine.chi_square.should be_close(7.384,0.001)
    end
    it "method df return degrees of freedom of model" do 
      @engine.df.should==5
    end
    it "method chi_square_null return X^2 of null model" do
      @engine.chi_square_null.should be_close(3725.0596,0.001)
    end
    it "method df_null return degrees of freedom for null model" do
      @engine.df_null.should==10
    end
    it "method goodness_of_fit return GFI" do
      pending('Not Implemented')
      @engine.goodness_of_fit.should be_close(0.99426, 0.0001)
    end
    it "method adjusted_goodness_of_fit return AGFI" do
      pending('Not Implemented')
      @engine.adjusted_goodness_of_fit.should be_close(0.98277, 0.0001)
    end
    it "method rmsea and derivatives returns... RMSEA!" do
      @engine.rmsea.should be_close(0.030911, 0.0001)
    end
    it "method rmsea_alpha and rmsea_confidence_interval returns alpa and ci for RMSEA" do 
      pending('Not Implemented other methods')
      @engine.rmsea_alpha.should==0.90
      @engine.rmsea_confidence_interval[0].should be_nil
      @engine.rmsea_confidence_interval[1].should be_close(0.074569, 0.0001)
    end
    it "method nfi returns NFI" do
      @engine.nfi.should be_close(0.99802, 0.0001)
    end
    it "method nnfi returns NNFI" do
     
      @engine.nnfi.should be_close(0.99872, 0.0001)
    end
    it "method cfi returns CFI" do
      @engine.cfi.should be_close(0.99936, 0.0001)
    end
    it "method srmr returns SRMR" do
      pending('Not Implemented')
      @engine.srmr.should be_close(0.0032191, 0.0001)
    end
    it "method bic returns BIC" do
      pending("Not well implemented")
      @engine.bic.should be_close(-23.689, 0.001)
    end
    it "method iterations returns the number of iterations" do
      pending('Not Implemented')
      @engine.iterations.should==30
    end
    
    # G to x1 0.397152 0.0155826 25.4869 0        x1 <--- G 
    # G to x2 0.503662 0.0182767 27.5576 0        x2 <--- G 
    # G to x3 0.577242 0.0204998 28.1584 0        x3 <--- G 
    # G to x4 0.702774 0.0240755 29.1905 0        x4 <--- G 
    # G to x5 0.796251 0.0267431 29.7741 0        x5 <--- G 
    # var x1  0.040814 0.0028254 14.4455 0        x1 <--> x1
    # var x2  0.038020 0.0028185 13.4893 0        x2 <--> x2
    # var x3  0.040827 0.0031653 12.8986 0        x3 <--> x3
    # var x4  0.039387 0.0034217 11.5110 0        x4 <--> x4
    # var x5  0.036287 0.0036916  9.8296 0        x5 <--> x5    
    it "coefficients returns a hash of Parameter estimates" do
      coeffs=@engine.coefficients
      coeffs[['G','x1']][:estimate].should be_close(0.397152, 0.00001)
      coeffs[['G','x1']][:se].should be_close(0.0155826, 0.0001)
      coeffs[['G','x1']][:label].should=='G to x1'
     # pending('Not Implemented z')
     coeffs[['G','x1']][:z].should be_close(coeffs[['G','x1']][:estimate].quo(coeffs[['G','x1']][:se]), 0.0001)
    end
    
    
  end
  if(false)
  describe "using raw data" do 
    before(:each) do
      @model=Statsample::SEM::Model.new do |m|
        m.manifests @ds.fields
        m.latents %w{G}
        m.path :from=>m.latents, :to=>m.manifests
        m.path :from=>m.manifests
        m.path :from=>m.latents, :free=>false, :values=>1.0
        m.path :from=>"one", :to=>m.manifests
        m.data_from_dataset(@ds)
      end
      @engine=Statsample::SEM::OpenMxEngine.new(@model)
    end
    it "should generate a valid r query" do
      @engine.r_query.size.should>=0
    end
    it "should have valid std coeffs" do
      pending()
      sc=@engine.standarized_coefficients
      sc.should be_instance_of Hash
      sc[['G','x1']][:estimate].should be_close(0.891309,0.0001)
    
    end
    it "should compute and return well formed response" do
      lambda{@engine.compute}.should_not raise_error
      @engine.r_summary.should be_instance_of (Array)
    end
    it "should return a valid graphviz definition for model" do
      @engine.graphviz.should match "digraph"
    end
    
    
  end
  end
  
end



example/wheaton.rb
$:.unshift(File.dirname(__FILE__)+"/../lib")
require 'statsample'
require 'statsample/sem'
matrix= Matrix[
[11.834,  6.947,    6.819,   4.783,  -3.839, -21.899],
[6.947,   9.364,    5.091,   5.028,  -3.889, -18.831], 
[6.819,   5.091,    12.532,  7.495,  -3.841, -21.748],
[4.783,   5.028,    7.495,   9.986,  -3.625, -18.775],
[-3.839,  -3.889,  -3.841,  -3.625,   9.610,  35.522],
[-21.899, -18.831, -21.748, -18.775, 35.522,  450.288]
]

matrix.extend Statsample::CovariateMatrix
matrix.fields=%w{Anomia67 Powerless67 Anomia71 Powerless71 Education SEI}

model=Statsample::SEM::Model.new(:name=>"Wheaton data") do |m|
  m.data_from_matrix(matrix, :cases=>932)
  m.latents=%w{Alienation67 Alienation71 SES}
  m.manifests=matrix.fields
  m.path :from=>"Alienation67", :to=>%w{Anomia67}, :free=>false, :values=>1.0
  m.path :from=>"Alienation67", :to=>%w{Powerless67 Alienation71}, :labels=>["lam1","beta"]
  m.path :from=>"Alienation71", :to=>%w{Anomia71}, :free=>false, :values=>1.0
  m.path :from=>"Alienation71", :to=>%w{Powerless71}, :labels=>["lam2"]
  m.path :from=>"SES", :to=>%w{Alienation67 Alienation71}, :labels=>%w{gam1 gam2}
  m.path :from=>"SES", :to=>%w{Education}, :free=>false, :values=>1.0
  m.path :from=>"SES", :to=>%w{SEI}, :labels=>["lam3"]
  m.path :from=>m.manifests, :arrows=>2, :labels=>%w{the11 the22 the33 the44 thd11 thd22}
  m.path :from=>m.latents, :arrows=>2, :labels=>%w{psi11 psi22 phi}
end
engine=Statsample::SEM::SemJFoxEngine.new(model, :name=>"Wheaton data")
engine.compute
puts engine.summary
model2=model.dup

model2.path :from=>"Anomia67", :to=>"Anomia71", :labels=>%w{the13}, :arrows=>2
engine2=Statsample::SEM::SemJFoxEngine.new(model2, :name=>"Wheaton data with the13")
engine2.compute
pp 1-Distribution::ChiSquare.cdf(engine2.chi_square, engine2.df.to_i)
puts engine2.summary


lib/statsample/sem.rb
require 'rserve'
require 'forwardable'
require 'statsample'
require 'statsample/rserve_extension'
require 'statsample/sem/model'
require 'statsample/sem/engine'
require 'statsample/sem/openmxengine'
require 'statsample/sem/semjfoxengine'

module Statsample
  class SEM
    extend Forwardable
    VERSION='0.1.0'
    attr_accessor :name
    attr_accessor :engine
    attr_reader :model
    attr_reader :engine_obj
    def_delegators :@model, :path, :data_from_matrix, :data_from_dataset, :manifests, :manifests=, :latents, :latents=, :data_variables, :data_variables=, :make_null
    def_delegators :@engine_obj, :summary, :chi_square, :df
    def initialize(opts=Hash.new, &block)
      default_opts={:name=>"SEM Analysis", :engine=>:sem}
      @opts=default_opts.merge(opts)
      @engine_obj=nil
      @engine = @opts.delete :engine
      @model=Statsample::SEM::Model.new(:name=>@name)
      if block
        block.arity<1 ? instance_eval(&block) : block.call(self)
      end
    end
    
    def name=(v)
        @opts[:name]=v
    end
    def name
      @opts[:name]
    end
    def compute
      @engine_obj=case @engine
      when :openmx
        OpenMxEngine.new(@model, @opts)
      when :sem
        SemJFoxEngine.new(@model, @opts)
      end
      @engine_obj.compute
      @engine_obj
    end
    def engine_obj
      @engine_obj||=compute
    end
    def r_summary
      engine_obj.r_summary
    end
  end
end


lib/statsample/sem/model.rb
module Statsample
  class SEM
    class Model
      include Summarizable
      attr_accessor :name
      # Type of data used. Could be +:covariance+, +:correlation+ or +:raw+
      attr_reader :data_type
      # Covariance/correlation matrix
      attr_reader :matrix
      # Raw data on a dataset
      attr_reader :ds
      # Number of cases
      attr_accessor :cases
      # Optional array of mean for use when data is matrix based
      attr_accessor :means
      # Name of data variables
      attr_reader :data_variables
      
      attr_reader :paths
      def initialize(opts=Hash.new, &block)
        raise ArgumentError,"opts should be a Hash" if !opts.is_a? Hash
        default_opts={:name=>_("SEM Model")}
        @opts=default_opts.merge opts
        @paths=Hash.new
        @manifests=nil
        @latents=nil
        if block
          block.arity<1 ? self.instance_eval(&block) : block.call(self)
        end
      end
      def dup
        mod=Model.new(@opts)
        if @data_type==:raw
          mod.data_from_dataset(@ds)
        else
          mod.data_from_matrix(@matrix,:type=>@data_type,:cases=>@cases, :means=>@means) 
        end
        mod.latents=latents
        mod.manifests=manifests
        if @paths
          @paths.each_pair do |k,v|
            mod.add_raw_path(v[:from],v[:to],v[:arrow],v[:label],v[:free], v[:value])
          end
        end
        mod
      end
      def make_null
        # Get current variables
        vars=variables_on_paths
        @paths.clear
        vars.each do |v|
          if @data_variables.include? v
            free=true
            value=nil
          else
            free=false
            value=1.0
          end
          add_raw_path(v,v,2,"var #{v}",free,value)
        end
      end
      def delete_path(f1,f2)
        
        @paths.delete([f1,f2].sort)
      end
      # True if model have enough information to process it
      def complete?
        !@data_type.nil? and !@manifests.nil? and !@latents.nil? and @paths.size>0
      end
      def add_raw_path(f1,f2,arrows,label,free,value)
        @paths[[f1,f2].sort]={ :from=>f1, :to=>f2, :arrow=>arrows, :label=>label, :free=>free, :value=>value}
      end
      def add_path(f1,f2,arrows,free,values,labels,i)
        arrow_s = (arrows==1) ? "to":"cov"
        raise "Path repeated : #{f1},#{f2}" if @paths.has_key? [[f1,f2].sort]
        label= (labels.nil? or !labels.respond_to?:[] or labels[i].nil?) ? "#{f1} #{arrow_s} #{f2}" : labels[i]
        
        free_v = (free.is_a? Array) ? free[i] : free 
        if values.is_a? Array
          value= values[i].nil? ? nil : values[i]
        elsif values.is_a? Numeric
          value=values
        end
          
        value = nil if free_v
        
        @paths[[f1,f2].sort]={ :from=>f1, :to=>f2, :arrow=>arrows, :label=>label, :free=>free_v, :value=>value}
      end
      # Set one or more paths. Based on OpenMx mxPath method.
      # 
      # ==Options:
      # * +:from+   : String or Array. sources of new paths
      # * +:to+     : String or Array. sinks of new paths
      # * +:all+    : bool. If you, connect all sources to all sinks. If false,
      # connect one-on-one sources to sinks if both are arrays, one source to
      # many sinks if +:from+ is a String and +:to+ is an Array
      # * +:arrows+ : 1 for regression, 2 for variance-covariance. See rules
      # for specific automatic setting
      # * +:free+   : Indicates whether paths are free or fixed. By default, true
      # * +:values+ : The starting values of the parameters
      # * +:labels+ : The names of paths
      # ==Rules
      # * from : variance-> from and to equal, label equal to "s^2 NAME_OF_FIELD", arrows=2, free=>true
      # * from and to: regression -> label equal to "FROM->TO", arrows=1, free=>true
      # * from, to, arrows -> label equal to "FROM->TO" if arrows=1,
      # "FROM<->TO" if arrows=2, free=>true
      # * free=false -> requires values for each from - to value
      
      def path(opts)
        raise "Requires at least :from option" unless opts.has_key? :from
        
        free=true
        all=false
        
        from=opts[:from]
        to=opts[:to]
        
        all=opts[:all] if opts.has_key? :all
        free=opts[:free] if opts.has_key? :free

        labels=opts[:labels]
        arrows=opts[:arrows]
        
        values=opts[:values]
        from=[from] if from.is_a? String
        to||=from
        to=[to] if to.is_a? String
        
        if from==to # variances
          arrows||=2
          labels_2=Array.new(from.size)
          
          from.each_with_index do |f,i|
            labels_2[i]=(labels.nil? or !labels.respond_to?:[] or labels[i].nil?) ? "var #{f}" : labels[i]
          end
          
          from.each_with_index do |f,i|
            add_path(f,f,arrows,free,values, labels_2,i)
          end
        else # regression and covariances
          arrows||=1
          i=0
          all=true if from.size==1 or to.size==1
          if all
            from.each do |f1|
              to.each do |f2|
                add_path(f1,f2,arrows,free,values,labels,i)
                i+=1 
              end
            end
          else
            raise ":from and :to should be the same size" if from.size!=to.size
            from.size.times.each do |i|
              add_path(from[i],to[i],arrows,free,values,labels,i)
            end
            
          end
          
        end
        
        
      end
      def data_variables=(v)
        @data_variables=v.map {|i| i.to_s}
      end
      # Total number of variables
      def variables_on_paths
        out=Array.new
        @paths.keys.each {|k|
          out << k[0] unless out.include? k[0]
          out << k[1] unless out.include? k[1]
        }
        out
      end
      # Number of variables on model
      def k
        variables_on_paths.size
      end
      # Latents will be any variable set on a path not present
      # on @variables
      def infer_latents
        variables_on_paths-@data_variables
      end
      # Latens will be any variable set on path present on @variables
      def infer_manifests
        variables_on_paths & @data_variables
      end
      def get_label(key)
        @paths[key][:label]
      end
      def latents(*argv)
        if argv.size==0
          if @latents.nil? and @paths.size>0 and !@data_type.nil? 
            @latents=infer_latents
          end
          @latents
        elsif argv[0].is_a? Array 
          @latents=argv[0]
        else
          @latents=[argv]
        end
      end
      def latents=(argv)
        @latents=argv
      end
      
      def manifests(*argv)
        if argv.size==0
          if @manifests.nil? and @paths.size>0 and !@data_type.nil? 
            @manifests=infer_manifests
          end
          @manifests
        elsif argv[0].is_a? Array 
          @manifests=argv[0]
        else
          @manifests=[argv]
        end
      end
      def manifests=(argv)
        @manifests=argv
      end
      def data_from_dataset(ds)
        @data_type=:raw
        @ds=ds
        @data_variables=@ds.fields
      end
      
      def data_from_matrix(matrix,opts=Hash.new)
        type = opts[:type]
        type||=(matrix.respond_to? :type) ? matrix.type : :covariance
        data_variables = opts[:data_variables]
        cases = opts[:cases]
        means = opts[:means]
        raise "You should set number of cases" if cases.nil?
        
        @data_type= (type==:covariance) ? :covariance : :correlation
        @matrix=matrix
        @cases=cases
        @means=means
        if data_variables.nil?
          if @matrix.respond_to? :fields
            @data_variables=@matrix.fields 
          elsif @data_variables.nil?
            @data_variables=@matrix.row_size.times.map {|i| "VAR#{i+1}"}
            
          end
        else
          @data_variables=variable_names
        end
      end
    end
  end
end

example/wheaton_alternate.rb
$:.unshift(File.dirname(__FILE__)+"/../lib")
require 'statsample'
require 'statsample/sem'
matrix= Matrix[
[11.834,  6.947,    6.819,   4.783,  -3.839, -21.899],
[6.947,   9.364,    5.091,   5.028,  -3.889, -18.831], 
[6.819,   5.091,    12.532,  7.495,  -3.841, -21.748],
[4.783,   5.028,    7.495,   9.986,  -3.625, -18.775],
[-3.839,  -3.889,  -3.841,  -3.625,   9.610,  35.522],
[-21.899, -18.831, -21.748, -18.775, 35.522,  450.288]
]

matrix.extend Statsample::CovariateMatrix
matrix.fields=%w{Anomia67 Powerless67 Anomia71 Powerless71 Education SEI}

model=Statsample::SEM::Model.new(:name=>"Wheaton data") do |m|
  m.data_from_matrix(matrix, :cases=>932)
  m.latents=%w{Alienation67 Alienation71 SES}
  m.manifests=matrix.fields
  m.path :from=>"Alienation67", :to=>%w{Anomia67 Powerless67 Alienation71}, :labels=>%w{aa1 aa2 aa3}
  m.path :from=>"Alienation71", :to=>%w{Anomia71 Powerless71}, :labels=>%w{bb1 bb2}
  m.path :from=>"SES", :to=>%w{Alienation67 Alienation71 Education SEI}, :labels=>%w{gam1 gam2 gam3 gam4}
  
  m.path :from=>m.manifests, :arrows=>2, :labels=>%w{the11 the22 the33 the44 thd11 thd22}
  m.path :from=>m.latents, :arrows=>2, :free=>false, :values=>1.0
  m.path :from=>"Anomia67", :to=>"Anomia71", :labels=>%w{the13}, :arrows=>2

end
engine=Statsample::SEM::SemJFoxEngine.new(model, :name=>"Wheaton data, alternate RAM")
engine.compute
puts engine.summary
model2=model.dup

lib/statsample/sem/engine.rb
module Statsample
  class SEM
    # abstract engine for SEM classes
    class Engine
      def common_summary(s)
        s.text _("Manifests: %s") % @model.manifests.join(", ")
        s.text _("Latents  : %s") % @model.latents.join(", ")
        s.text "Chi-square: %0.3f (d.f=%d), p = %0.5f " % [chi_square, df, (1-Distribution::ChiSquare.cdf(chi_square, df))]
        s.table(:name=>_("Parameter estimation"),:header=>[_("From"), _("To"), _("Label"),  _("estimate"),_("se"), _("z")]) do |t|
          @model.paths.sort.each do |v|
            
            f1,f2 = v[0][0],v[0][1]
            key=v[0]

            if v[1][:free]
              val=coefficients[key]
              raise "Doesn't have coefficient #{key}" if val.nil?
              label=v[1][:label]
              estimate = val[:estimate].nil? ? "?" : "%0.5f" % val[:estimate]
              se=val[:se].nil? ? "?" : ("%0.5f" % val[:se])
              
              z=(val[:z].nil? or val[:p].nil?) ? "?" : ("%0.3f%s(%0.2f)" % [val[:z], val[:z].abs>=1.96 ? "*":"", val[:p]]) 
            else
              label=_("%s (Fixed)") % v[1][:label]
              estimate=v[1][:value]
              se="--"
              z="--"
            end
            t.row [f1,f2, label, estimate, se, z] 
          end
        end
        
        def summary_standarized_coefficients_table(s)
        s.table(:name=>_("Standarized coefficients"),:header=>[_("From"), _("To"), _("Label"),  _("estimate")]) do |t|
          @model.paths.sort.each do |v|
            f1,f2 = v[0][0],v[0][1]
            key=v[0]
            if v[1][:free]
              val=standarized_coefficients[key]
              label=v[1][:label]
              estimate=val[:estimate].nil? ? "?" : "%0.5f" % val[:estimate]
            else
              label=_("%s (Fixed)") % v[1][:label]
              estimate=v[1][:value]
            end
            t.row [f1,f2, label, estimate] 
          end
        end
      end
        
      end
    end
  end
end

lib/statsample/sem/openmxengine.rb
require 'rserve'
require 'tempfile'
module Statsample
  class SEM
    class OpenMxEngine < Engine
      include Summarizable
      attr_accessor :summarizable
      attr_accessor :name
      attr_reader :model
      def initialize(model,opts=Hash.new)
        @model=model
        defaults = {
          :name=>_("SEM analysis using OpenMx"),
          :summary_standarized_coefficients=>false
        }
        @opts=defaults.merge opts
        @name=@opts[:name]
      end
      def r
        @r||=Rserve::Connection.new
      end
      def r_mxpaths
        @model.paths.values.map {|path|
          value= path[:value] ? ", values = #{path[:value]}":""
          label= path[:label] ? ", labels = \"#{path[:label]}\"" : ""  
          "mxPath(from=c(\"#{path[:from]}\"), to=c(\"#{path[:to]}\"), arrows = #{path[:arrow]}, free = #{path[:free] ? "TRUE" : "FALSE"} #{value} #{label})"
        }.join(",\n")
      end
      def r_mxdata
        type=case @model.data_type
          when :raw
            'raw'
          when :covariance
            'cov'
          when :correlation
            'cor'
        end
        means=(@model.data_type!=:raw and !@model.means.nil?) ? ", means = d_means " : ""
        num=(@model.data_type!=:raw) ? ", numObs = #{@model.cases} " : ""
        
        "mxData(observed=data, type='#{type}' #{means} #{num})"
      end
        
      def r_query
        <<-EOF
library(OpenMx);
factorModel <- mxModel(
name="#{name}",
type="RAM",
manifestVars = manifests,
latentVars = latents,
#{r_mxpaths}, 
#{r_mxdata}
);
factorFit<-mxRun(factorModel);
rm(data,manifests,latents,d_means);
        EOF
        #p r.eval('factorFit').to_ruby
      end
      
      def compute
        raise "Insuficient information" unless @model.complete?
        r.assign 'data', @model.data_type==:raw ? @model.ds : @model.matrix
        if @model.matrix
          r.assign 'vn', @model.data_variables
          # We should assing names to fields on matrix
          r.void_eval('dimnames(data)<-list(vn,vn)')
        end
        r.assign 'manifests',@model.manifests
        r.assign 'latents', @model.latents
        r.assign 'd_means',@model.means unless @model.means.nil?
        r.void_eval r_query
        @r_summary=@r.eval('summary(factorFit)').to_ruby
        
      end
      def graphviz
        compute if @r_summary.nil?
        tf=Tempfile.new('model.dot')
        r.void_eval("omxGraphviz(factorModel,'#{tf.path}')")
#        tf.close
#        tf.open
        tf.read
      end
      def r_summary
        @r_summary||=compute
      end
      def chi_square
        r_summary['Chi']
      end
      def df
        r_summary['degreesOfFreedom']
      end
      def chi_square_null
        null_model.r_summary['Chi']
      end
      def df_null
        null_model.r_summary['degreesOfFreedom']
      end

      def null_model
        @null_model||=compute_null_model
      end
      def compute_null_model #:nodoc:
        nm=@model.dup
        nm.make_null
        (self.class).new(nm,@opts)
      end
      def rmsea
        r_summary['RMSEA']
      end
       # [χ2(Null Model) - χ2(Proposed Model)]/ [χ2(Null Model)]
      def nfi
        (chi_square_null-chi_square).quo(chi_square_null)
      end
      
      # [χ2/df(Null Model) - χ2/df(Proposed Model)]/[χ2/df(Null Model) - 1]
      def nnfi
        (chi_square_null.quo(df_null) - chi_square.quo(df)).quo(chi_square_null.quo(df_null)-1)
      end
      def cfi
        d_null=chi_square_null-df_null
        ((d_null)-(chi_square-df)).quo(d_null)
      end
      def bic
        raise "Not well implemented"
        k=@model.k
        ln_n=Math.log(@model.cases)
        chi_square+((k*(k-1).quo(2)) - df)*ln_n
      end
      def coefficients
        @coefficients||=compute_coefficients
      end
      
      
      def standarized_coefficients
        if @std_coef.nil?
          @std_coef=model_with_correlation.coefficients
        end
      @std_coef
      end
      
      def model_with_correlation
        if @model_with_correlation.nil?
          if model.matrix.type==:covariance
            model2=model.dup
            model2.data_from_matrix( model.matrix.correlation, :cases=>model.cases )
            @model_with_correlation=(self.class).new(model2)
          else
            @model_with_correlation=self
          end
        end
        @model_with_correlation
      end
      
      
      def compute_coefficients #:nodoc:
        est=Hash.new
        coeffs=r_summary['parameters']
        # 0:name, 1:matrix, 2:row, 3:col, 4:estimate, 5:Std.error
        coeffs[0].each_with_index do |v,i|
          f1=coeffs[2][i]
          f2=coeffs[3][i]
          key=[f1,f2].sort
          z=coeffs[4][i].quo(coeffs[5][i])
          est[key]={:estimate=>coeffs[4][i], :se=>coeffs[5][i], :z=>z, :p=>Statsample::Test.p_using_cdf(Distribution::Normal.cdf(z)), :label=>v}
        end
        est
        
      end
      def report_building(g)
        g.section(:name=>@name) do |s|
          common_summary(s)
          summary_standarized_coefficients_table(s) if summary_standarized_coefficients
        end
      end

    end
  end
end


example/normal_vs_saturated.rb
$:.unshift(File.dirname(__FILE__)+"/../lib/")
require 'statsample'
require 'statsample/sem'
require 'matrix'


matrix=Matrix[[0.1985443,0.1999953,0.2311884,0.2783865,0.3155943],
[0.1999953,0.2916950,0.2924566,0.3515298,0.4019234],
[0.2311884,0.2924566,0.3740354,0.4061291,0.4573587],
[0.2783865,0.3515298,0.4061291,0.5332788,0.5610769],
[0.3155943,0.4019234,0.4573587,0.5610769,0.6703023]]


cases=500
sem1=Statsample::SEM.new(:name=>"SEM analysis") do |m|
    m.data_from_matrix(matrix,:cases=>cases)
    m.manifests m.data_variables
    m.latents %w{G}
    m.path :from=>m.latents, :to=>m.manifests
    m.path :from=>m.manifests
    m.path :from=>m.latents, :free=>false, :values=>1.0
    
end
sem1.compute
puts sem1.summary

sem2=sem1.dup
sem2.make_null

sem2.name="Null Model"

sem2.compute

puts sem2.summary

